### PR TITLE
[WIP] Scenarios Generators

### PR DESCRIPTION
**Pull request by TLESORT**

### CODE REVIEW

File: continuum/generators/scenarios_generator.py [hunk 1]
```diff
@@ -0,0 +1,63 @@
+import abc
+from continuum.scenarios import _BaseScenario, create_subscenario
+import torch
```
Comment (arthurdouillard) [inline continuum/generators/scenarios_generator.py:3]: Can you sort a bit the import please?

BUILTIN PACKAGES
empty line
PYPI PACKAGES
empty line
INTERNAL IMPORTS

File: continuum/generators/scenarios_generator.py [hunk 2]
```diff
@@ -0,0 +1,63 @@
...
+    DO NOT INSTANTIATE THIS CLASS.
+
+    :param scenario: A Continuum scenarios
+    """
+
+    def __init__(
+            self,
+            scenario: _BaseScenario
+    ) -> None:
+        self.base_scenario = scenario
```
Comment (arthurdouillard) [inline continuum/generators/scenarios_generator.py:19]: Do we accept an instantiated scenario or the base class? Because in some situation (like sorting per color) we may want to mix the data of different task ids

File: continuum/generators/scenarios_generator.py [hunk 3]
```diff
@@ -0,0 +1,63 @@
...
+        self.nb_generator.manual_seed(seed)
+
+        # generate a random task order
+        task_order = torch.randperm(self.base_scenario.nb_tasks, generator=self.nb_generator)
+        if nb_tasks is None:
+            nb_tasks = self.base_scenario.nb_tasks
+        return task_order[:nb_tasks]
+
+    def sample(self, seed: int = None, nb_tasks: int = None) -> _BaseScenario:
+        # seed the generator
```
Comment (arthurdouillard) [inline continuum/generators/scenarios_generator.py:53]: It would be simpler if you remove those two lines but add a `if seed is not None` line 44

File: continuum/generators/scenarios_generator.py [hunk 4]
```diff
@@ -0,0 +1,63 @@
...
+        if seed is None:
+            seed = np.random.randint(10000)
+
+        # generate a random task order
+        task_order = self.get_class_order(seed, nb_tasks)
+
+        subscenario = create_subscenario(self.base_scenario, task_order[:nb_tasks])
+
+        # return subscenario
```
Comment (arthurdouillard) [inline continuum/generators/scenarios_generator.py:62]: I'm not sure this comment is really necessary ^^

File: continuum/generators/scenarios_generator.py (hunk 2 above)
Comment (TLESORT) [reply continuum/generators/scenarios_generator.py:19]: An instantiated scenario, should I make an union all types of scenarios for that?

File: continuum/generators/scenarios_generator.py (hunk 3 above)
Comment (TLESORT) [reply continuum/generators/scenarios_generator.py:53]: En fait get_class_order peut etre utilisé par un program exterieur pour recuperer le task order, du coup je voulais forcer le user a donner une seed pour cette fonction.

File: continuum/generators/scenarios_generator.py (hunk 1 above)
Comment (TLESORT) [reply continuum/generators/scenarios_generator.py:3]: only abc is builtin package right?

Comment (arthurdouillard) [reply continuum/generators/scenarios_generator.py:3]: yep

File: continuum/generators/scenarios_generator.py (hunk 2 above)
Comment (arthurdouillard) [reply continuum/generators/scenarios_generator.py:21]: I'm not talking directly about the Typing, but rather the user experience: it will always be an instanciated scenario? or it could be a scenario + a dataset?

Because I'm not sure how you could do some split (like color) in the former

Comment (TLESORT) [reply continuum/generators/scenarios_generator.py:21]: maybe scenario + dataset is more general, and makes possible more types of generator

Comment (arthurdouillard) [reply continuum/generators/scenarios_generator.py:21]: I agree

File: continuum/generators/scenarios_generator.py (hunk 3 above)
Comment (arthurdouillard) [reply continuum/generators/scenarios_generator.py:55]: Ok why not

File: continuum/generators/scenarios_generator.py (hunk 2 above)
Comment (TLESORT) [reply continuum/generators/scenarios_generator.py:21]: For ` ClassOrderGenerator `  and ` `TaskOrderGenerator `  it is a instantiate scenario to not have to forward all the parameters. But other generators might be used differently.

File: continuum/scenarios/hashed.py [hunk 5]
```diff
@@ -0,0 +1,124 @@
...
+
+        example_per_tasks = nb_examples // nb_tasks
+        perfect_balance_task_ids = np.arange(nb_tasks).repeat(example_per_tasks)
+        task_ids[:len(perfect_balance_task_ids)] = perfect_balance_task_ids
+
+        # examples from len(perfect_balance_task_ids) to len(task_ids) are affected to last tasks
+
+        return task_ids
+
+    def generate_task_ids(self, cl_dataset, hash_name, nb_tasks):
```
[the remark this answers is not in the document]
Comment (TLESORT) [reply continuum/scenarios/hashed.py:102]: I did something related to automatically set the number of task or create more "hash-coherent" tasks

**Comment by TLESORT** [comment]
> I think you should convert the list to str before. Here the str conversion in the key can be done up to nlog(n) times instead of n times.
The str conversion was the first attempt to order hashes, but I think that the ordering based on str is wrong. I looking for a better way to order hashes....

**Comment by TLESORT** [comment]
> 
> 
> Does a distance (l1, l2, whatever) makes sense in the hash space? I'm not sure it's isometric.

The distance is computed as here: https://github.com/JohannesBuchner/imagehash/blob/master/imagehash.py#L97-L102
I do not know if using hamming distance in str space would make a homogeneous distance. I am investigating since it would be more convenient to use.
